ansible_base/jwt_consumer/common/auth.py

```diff
@@ -40,35 +48,96 @@ def parse_jwt_token(self, request):
+            # Since we got a new key, lets go ahead and try to validate the token again.
+            # If it fails this time we can just raise whatever
+            validated_body = self.validate_token(token, new_decryption_key)
+
+        # Let's see if we have the same user info in the cache already
+        is_cached, user_defaults = self.check_user_in_cache(validated_body)
+
         user_model = get_user_model()
-        user, created = user_model.objects.update_or_create(
-            username=validated_body["sub"],
-            defaults={
-                "first_name": validated_body["first_name"],
-                "last_name": validated_body["last_name"],
-                "email": validated_body["email"],
-                "is_superuser": validated_body["is_superuser"],
-            },
-        )
+        user = None
+        if is_cached:
+            try:
+                user = user_model.objects.get(username=validated_body['sub'])
+            except user_model.DoesNotExist:
+                # ooofff... I'm sorry, you user was in the cache but deleted from the database?
+                # but now you have to pay the price to continue logging in
+                pass
+
+        if not user:
+            # Either the user wasn't cached or the requested user was not in the DB so we need to make a new one
+            user, created = user_model.objects.update_or_create(
+                username=validated_body["sub"],
+                defaults=user_defaults,
+            )
+            if created:
+                logger.warn(f"New user {user.username} created from JWT auth")
+
         setattr(user, "resource_api_actions", validated_body.get("resource_api_actions", None))
-        if created:
-            logger.warn(f"New user {user.username} created from JWT auth")
-        else:
-            logger.info(f"User {user.username} authenticated from JWT auth")
 
+        logger.info(f"User {user.username} authenticated from JWT auth")
         return user, validated_body
 
     def log_and_raise(self, conditional_translate_object, expand_values={}):
         logger.error(conditional_translate_object.not_translated() % expand_values)
         raise AuthenticationFailed(conditional_translate_object.translated() % expand_values)
 
-    def get_decryption_key_from_url(self, url, timeout, validate_certs):
+    def check_user_in_cache(self, validated_body: dict) -> Tuple[bool, dict]:
+        # These are the defaults which will get passed to the user creation and what we expect in the cache
+        expected_cache_value = {
+            "first_name": validated_body["first_name"],
+            "last_name": validated_body["last_name"],
+            "email": validated_body["email"],
+            "is_superuser": validated_body["is_superuser"],
+        }
+        cached_user = cache.get(validated_body["sub"], None)
```

Comment:
It might be better to use the JWT signature as the cache key here with the user ID as the value. That way you can validate the incoming token by just taking the signature and checking if it is in the cache. If it is, you can assume that the token has already been validated and trust it without having to recheck the signature.

Reply:
I thought about this but a JWT will be reissued if it expires which, I think, will change the signature. Going by the fields gets us around that while still causing processing if they subject of the key actually changes.

Reply:
ah, so this is just to determine if the user details need to be updated. Got it. I had in mind that we'd be caching the whole token and redoing the check every time the token expires.